File: backend_docker_large/backend/routes/files.js
// ========================================
// SECURE FILE PRINTING SYSTEM - FILE ROUTES
// POST /api/upload - Upload encrypted file
// GET /api/files - List all uploaded files
// GET /api/print/:id - Download file for printing
// POST /api/delete/:id - Delete file after printing
// ========================================

const express = require("express");
const router = express.Router();
const multer = require("multer");
const { v4: uuidv4 } = require("uuid");
const db = require("../database");
const { verifyToken, verifyRole } = require("../middleware/auth");

// Configure multer for file uploads (store in memory, then save to DB)
const upload = multer({
  storage: multer.memoryStorage(),
  limits: {
    fileSize: 500 * 1024 * 1024, // 500MB max
  },
});

// ========================================
// 1. POST /api/upload
// Upload encrypted file from mobile app
// ========================================
router.post("/upload", verifyToken, upload.single("file"), async (req, res) => {
  try {
    // Validate request
    if (!req.file) {
      return res.status(400).json({ error: "No file provided" });
    }

    // File size validation (double check on server side)
    const maxFileSize = 500 * 1024 * 1024; // 500MB
    if (req.file.size > maxFileSize) {
      return res.status(413).json({
        error: "File too large",
        max_size_mb: maxFileSize / (1024 * 1024),
        file_size_mb: (req.file.size / (1024 * 1024)).toFixed(2),
      });
    }

    if (!req.body.file_name) {
      return res.status(400).json({ error: "file_name is required" });
    }

    if (!req.body.file_name.match(/^[\w\s.\-]+$/)) {
      return res.status(400).json({
        error: "file_name contains invalid characters",
        allowed: "alphanumeric, spaces, dots, hyphens",
      });
    }

    if (!req.body.iv_vector) {
      return res.status(400).json({ error: "iv_vector is required" });
    }

    if (!req.body.auth_tag) {
      return res.status(400).json({ error: "auth_tag is required" });
    }

    // Ensure owner_id is provided (owner who will print)
    const ownerId = req.body.owner_id;
    if (!ownerId) {
      return res.status(400).json({ error: "owner_id is required" });
    }

    // Generate unique file ID
    const fileId = uuidv4();

    if (!req.body.encrypted_symmetric_key) {
      return res.status(400).json({ error: "encrypted_symmetric_key is required" });
    }

    // Convert base64 strings to Buffer
    const ivBuffer = Buffer.from(req.body.iv_vector, "base64");
    const authTagBuffer = Buffer.from(req.body.auth_tag, "base64");
    const encryptedKeyBuffer = Buffer.from(req.body.encrypted_symmetric_key, "base64");

    // Try to insert into database (include user_id and owner_id)
    try {
      const query = `
      INSERT INTO files (
        id,
        user_id,
        owner_id,
        file_name,
        encrypted_file_data,
        file_size_bytes,
        file_mime_type,
        iv_vector,
        auth_tag,
        encrypted_symmetric_key,
        created_at,
        is_deleted
      ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, NOW(), false)
      RETURNING id, file_name, file_size_bytes, created_at
    `;

      const result = await db.query(query, [
        fileId,
        req.user && req.user.sub ? req.user.sub : null, // user_id from JWT
        ownerId,
        req.body.file_name,
        req.file.buffer,
        req.file.size,
        req.file.mimetype || "application/octet-stream",
        ivBuffer,
        authTagBuffer,
        encryptedKeyBuffer,
      ]);

      const uploadedFile = result.rows[0];

      // Log successful upload with prominent formatting
      console.log('\n' + '='.repeat(60));
      console.log('🎉 FILE RECEIVED FROM MOBILE APP 🎉');
      console.log('='.repeat(60));
      console.log(`✅ File ID: ${fileId}`);
      console.log(`   Name: ${uploadedFile.file_name}`);
      console.log(`   Size: ${(uploadedFile.file_size_bytes / 1024 / 1024).toFixed(2)} MB`);
      console.log(`   Owner ID: ${ownerId}`);
      console.log(`   Uploaded at: ${uploadedFile.created_at.toISOString()}`);
      console.log('='.repeat(60) + '\n');

      res.status(201).json({
        success: true,
        file_id: uploadedFile.id,
        file_name: uploadedFile.file_name,
        file_size_bytes: uploadedFile.file_size_bytes,
        uploaded_at: uploadedFile.created_at.toISOString(),
        message: "File uploaded successfully. Share the file_id with the owner.",
      });
    } catch (dbError) {
      // If DB is unavailable (common in local dev), fallback to saving on disk
      console.error('DB insert error, falling back to disk storage:', dbError && dbError.message ? dbError.message : dbError);

      const fs = require('fs');
      const path = require('path');
      const uploadsDir = path.join(__dirname, '..', 'uploads');
      if (!fs.existsSync(uploadsDir)) {
        fs.mkdirSync(uploadsDir, { recursive: true });
      }

      const filePath = path.join(uploadsDir, `${fileId}.bin`);
      const metaPath = path.join(uploadsDir, `${fileId}.json`);

      try {
        fs.writeFileSync(filePath, req.file.buffer);
        const meta = {
          id: fileId,
          user_id: req.user && req.user.sub ? req.user.sub : null,
          owner_id: ownerId,
          file_name: req.body.file_name,
          file_size_bytes: req.file.size,
          file_mime_type: req.file.mimetype || 'application/octet-stream',
          iv_vector: req.body.iv_vector,
          auth_tag: req.body.auth_tag,
          created_at: new Date().toISOString(),
        };
        fs.writeFileSync(metaPath, JSON.stringify(meta, null, 2));

        // Log with prominent formatting (fallback mode)
        console.log('\n' + '='.repeat(60));
        console.log('🎉 FILE RECEIVED FROM MOBILE APP (DISK FALLBACK) 🎉');
        console.log('='.repeat(60));
        console.log(`✅ File ID: ${fileId}`);
        console.log(`   Name: ${req.body.file_name}`);
        console.log(`   Size: ${(req.file.size / 1024 / 1024).toFixed(2)} MB`);
        console.log(`   Owner ID: ${ownerId}`);
        console.log(`   Saved to: ${filePath}`);
        console.log('='.repeat(60) + '\n');

        res.status(201).json({
          success: true,
          file_id: fileId,
          file_name: req.body.file_name,
          file_size_bytes: req.file.size,
          uploaded_at: meta.created_at,
          message: 'File stored locally (DB unavailable). Use admin tools to import into DB later.',
          fallback: true,
        });
        return;
      } catch (fsErr) {
        console.error('Fallback disk write failed:', fsErr);
        throw fsErr; // let outer catch handle
      }
    }
  } catch (error) {
    console.error("Upload error:", error);
    const isDev = process.env.NODE_ENV === "development";
    res.status(500).json({
      error: true,
      message: isDev ? error.message : "Failed to upload file",
      ...(isDev && { details: error.stack }),
    });
  }
});

// ========================================
// 2. GET /api/files
// List all uploaded files waiting to be printed
// ========================================
router.get("/files", verifyToken, async (req, res) => {
  try {
    // If requester is a normal user return their uploads; if owner, return files assigned to owner
    let result;
    if (req.user && req.user.role === "user") {
      const q = `
        SELECT id, file_name, file_size_bytes, created_at, is_printed, printed_at
        FROM files
        WHERE is_deleted = false AND user_id = $1
        ORDER BY created_at DESC
        LIMIT 100
      `;
      result = await db.query(q, [req.user.sub]);
    } else if (req.user && req.user.role === "owner") {
      const q = `
        SELECT id, file_name, file_size_bytes, created_at, is_printed, printed_at
        FROM files
        WHERE is_deleted = false AND owner_id = $1
        ORDER BY created_at DESC
        LIMIT 100
      `;
      result = await db.query(q, [req.user.sub]);
    } else {
      // default: return recent non-deleted files (admin-like)
      const q = `
        SELECT id, file_name, file_size_bytes, created_at, is_printed, printed_at
        FROM files
        WHERE is_deleted = false
        ORDER BY created_at DESC
        LIMIT 100
      `;
      result = await db.query(q);
    }

    const files = result.rows.map((row) => ({
      file_id: row.id,
      file_name: row.file_name,
      file_size_bytes: row.file_size_bytes,
      uploaded_at: row.created_at.toISOString(),
      is_printed: row.is_printed || false,
      printed_at: row.printed_at ? row.printed_at.toISOString() : null,
      status: row.is_printed ? "PRINTED_AND_DELETED" : "WAITING_TO_PRINT",
    }));

    console.log(`✅ Listed ${files.length} files`);

    res.json({
      success: true,
      count: files.length,
      files: files,
      message: `${files.length} file(s) waiting to be printed`,
    });
  } catch (error) {
    console.error("List files error:", error);
    res.status(500).json({
      error: true,
      message: "Failed to list files",
      details: error.message,
    });
  }
});

// ========================================
// 3. GET /api/print/:file_id
// Download encrypted file for printing
// Owner receives encrypted file + IV + Auth Tag
// Owner must decrypt client-side before printing
// ========================================
// Only owners can download encrypted files for printing
router.get(
  "/print/:file_id",
  verifyToken,
  verifyRole(["owner"]),
  async (req, res) => {
    try {
      const { file_id } = req.params;

      // Validate file_id format
      if (!file_id || file_id.length < 5) {
        return res.status(400).json({ error: "Invalid file_id" });
      }

      // Query database for encrypted file
      const query = `
      SELECT 
        id,
        file_name,
        encrypted_file_data,
        file_size_bytes,
        iv_vector,
        auth_tag,
        encrypted_symmetric_key,
        created_at,
        is_printed,
        owner_id
      FROM files
      WHERE id = $1 AND is_deleted = false
    `;

      const result = await db.query(query, [file_id]);

      if (result.rows.length === 0) {
        return res.status(404).json({
          error: true,
          message: "File not found or already deleted",
          file_id: file_id,
        });
      }

      const file = result.rows[0];

      // Ensure the requesting owner is assigned to this file
      if (req.user && req.user.role === "owner") {
        if (file.owner_id && file.owner_id !== req.user.sub) {
          return res.status(403).json({
            error: true,
            message: "Forbidden: not assigned to this file",
          });
        }
      }

      console.log(`✅ File downloaded for printing: ${file_id}`);
      console.log(`   Name: ${file.file_name}`);
      console.log(`   Size: ${file.file_size_bytes} bytes`);
      console.log(
        `   Status: ${file.is_printed ? "Already printed" : "Ready to print"}`
      );

      // Return encrypted file with IV and auth tag
      // Client must decrypt before printing
      res.json({
        success: true,
        file_id: file.id,
        file_name: file.file_name,
        file_size_bytes: file.file_size_bytes,
        uploaded_at: file.created_at.toISOString(),
        is_printed: file.is_printed || false,
        // Client receives encrypted data as base64
        encrypted_file_data: file.encrypted_file_data.toString("base64"),
        iv_vector: file.iv_vector.toString("base64"),
        auth_tag: file.auth_tag.toString("base64"),
        encrypted_symmetric_key: file.encrypted_symmetric_key.toString("base64"),
        // Instructions for client
        message: "Decrypt this file on your PC before printing",
        decryption_instructions: {
          step1: "Receive the encrypted_file_data, iv_vector, and auth_tag",
          step2: "You must have the decryption key (shared by uploader)",
          step3:
            "Call decryptFileAES256(encrypted_file_data, key, iv_vector, auth_tag)",
          step4: "Decryption happens ONLY in memory (never touches disk)",
          step5: "Send decrypted data to printer",
          step6: "Call DELETE /api/delete/{file_id} to auto-delete",
        },
      });
    } catch (error) {
      console.error("Print download error:", error);
      res.status(500).json({
        error: true,
        message: "Failed to download file for printing",
        details: error.message,
      });
    }
  }
);

// ========================================
// 4. POST /api/delete/:file_id
// Delete file after printing
// Owner calls this AFTER printing is complete
// File is permanently deleted from database
// ========================================
// Only owners can mark files as printed/deleted
router.post(
  "/delete/:file_id",
  verifyToken,
  verifyRole(["owner"]),
  async (req, res) => {
    try {
      const { file_id } = req.params;

      // Validate file_id format
      if (!file_id || file_id.length < 5) {
        return res.status(400).json({ error: "Invalid file_id" });
      }

      // Check if file exists
      const checkQuery = `
      SELECT id, file_name, is_deleted, owner_id
      FROM files
      WHERE id = $1
    `;

      const checkResult = await db.query(checkQuery, [file_id]);

      if (checkResult.rows.length === 0) {
        return res.status(404).json({
          error: true,
          message: "File not found",
          file_id: file_id,
        });
      }

      const file = checkResult.rows[0];

      // Ensure owner matches
      if (req.user && req.user.role === "owner") {
        if (file.owner_id && file.owner_id !== req.user.sub) {
          return res.status(403).json({
            error: true,
            message: "Forbidden: not assigned to this file",
          });
        }
      }

      if (file.is_deleted) {
        return res.status(400).json({
          error: true,
          message: "File already deleted",
          file_id: file_id,
        });
      }

      // Delete file (mark as deleted + set deleted_at timestamp)
      const deleteQuery = `
      UPDATE files
      SET 
        is_deleted = true,
        deleted_at = NOW(),
        is_printed = true,
        printed_at = NOW()
      WHERE id = $1
      RETURNING id, file_name, deleted_at
    `;

      const deleteResult = await db.query(deleteQuery, [file_id]);
      const deletedFile = deleteResult.rows[0];

      console.log(`✅ File deleted: ${file_id}`);
      console.log(`   Name: ${deletedFile.file_name}`);
      console.log(`   Deleted at: ${deletedFile.deleted_at}`);

      res.json({
        success: true,
        file_id: deletedFile.id,
        file_name: deletedFile.file_name,
        status: "DELETED",
        deleted_at: deletedFile.deleted_at.toISOString(),
        message: "File has been permanently deleted from server",
      });
    } catch (error) {
      console.error("Delete error:", error);
      res.status(500).json({
        error: true,
        message: "Failed to delete file",
        details: error.message,
      });
    }
  }
);

// ========================================
// EXPORT ROUTER
// ========================================

module.exports = router;

File: backend_flask_small/routes/files.py
from flask import Blueprint, request, jsonify, g
from db import get_db_connection, release_db_connection
from auth_utils import token_required
from sse_manager import sse_manager
import uuid
import base64
import datetime

files_bp = Blueprint('files', __name__)

@files_bp.route('/upload', methods=['POST'])
# @token_required  <-- Disabled for testing
def upload_file():
    if 'file' not in request.files:
        return jsonify({'error': 'No file provided'}), 400
    
    file = request.files['file']
    file_name = request.form.get('file_name')
    iv_vector_b64 = request.form.get('iv_vector')
    auth_tag_b64 = request.form.get('auth_tag')
    encrypted_key_b64 = request.form.get('encrypted_symmetric_key')
    owner_id = request.form.get('owner_id')

    if not all([file_name, iv_vector_b64, auth_tag_b64, encrypted_key_b64, owner_id]):
        return jsonify({'error': 'Missing required fields'}), 400

    # Convert Base64 to Bytes for DB storage
    try:
        iv_vector = base64.b64decode(iv_vector_b64)
        auth_tag = base64.b64decode(auth_tag_b64)
        encrypted_key = base64.b64decode(encrypted_key_b64)
        file_data = file.read()
        file_size = len(file_data)
        file_mime = file.content_type or 'application/octet-stream'
    except Exception as e:
        return jsonify({'error': f'Invalid encoding: {str(e)}'}), 400

    file_id = str(uuid.uuid4())
    # user_id = g.user['sub'] <-- Disabled for testing
    user_id = 'test-user-id-for-development' # Hardcoded for testing

    conn = get_db_connection()
    cursor = conn.cursor()

    # Resolve Owner ID (Mobile app sends Email, we need UUID)
    real_owner_id = owner_id
    try:
        cursor.execute("SELECT id FROM owners WHERE email = ?", (owner_id,))
        row = cursor.fetchone()
        if row:
            real_owner_id = row[0]
            print(f"Resolved owner email '{owner_id}' to UUID '{real_owner_id}'")
        else:
            # Check if it's already a UUID
            cursor.execute("SELECT id FROM owners WHERE id = ?", (owner_id,))
            if cursor.fetchone():
                real_owner_id = owner_id
            else:
                print(f"Warning: Owner '{owner_id}' not found in DB. Using as-is.")
    except Exception as e:
        print(f"Error resolving owner: {e}")

    try:
        created_at = datetime.datetime.utcnow().isoformat()
        cursor.execute(
            """INSERT INTO files (
                id, user_id, owner_id, file_name, encrypted_file_data, 
                file_size_bytes, file_mime_type, iv_vector, auth_tag, 
                encrypted_symmetric_key, created_at, is_deleted
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 0)""",
            (
                file_id, user_id, real_owner_id, file_name, file_data,
                file_size, file_mime, iv_vector, auth_tag, encrypted_key, created_at
            )
        )
        conn.commit()

        print(f"File uploaded: {file_id} ({file_size} bytes)")

        # Notify desktop client via SSE
        sse_manager.publish(owner_id, "new_file", {
            "file_id": file_id,
            "file_name": file_name,
            "file_size_bytes": file_size,
            "uploaded_at": created_at
        })

        return jsonify({
            'success': True,
            'file_id': file_id,
            'file_name': file_name,
            'file_size_bytes': file_size,
            'uploaded_at': created_at,
            'message': 'File uploaded successfully'
        }), 201

    except Exception as e:
        conn.rollback()
        print(f"Upload error: {e}")
        return jsonify({'error': True, 'message': 'Upload failed'}), 500
    finally:
        cursor.close()
        release_db_connection(conn)

@files_bp.route('/files', methods=['GET'])
@token_required
def list_files():
    user_id = g.user['sub']
    role = g.user['role']

    conn = get_db_connection()
    cursor = conn.cursor()

    try:
        if role == 'user':
            query = """SELECT id, file_name, file_size_bytes, created_at, is_printed, printed_at 
                       FROM files WHERE is_deleted = 0 AND user_id = ? ORDER BY created_at DESC LIMIT 100"""
            cursor.execute(query, (user_id,))
        elif role == 'owner':
            query = """SELECT id, file_name, file_size_bytes, created_at, is_printed, printed_at 
                       FROM files WHERE is_deleted = 0 AND owner_id = ? ORDER BY created_at DESC LIMIT 100"""
            cursor.execute(query, (user_id,))
        else:
            return jsonify({'error': 'Invalid role'}), 403

        files = []
        for row in cursor.fetchall():
            files.append({
                'file_id': row[0],
                'file_name': row[1],
                'file_size_bytes': row[2],
                'uploaded_at': row[3],
                'is_printed': row[4],
                'printed_at': row[5],
                'status': 'PRINTED_AND_DELETED' if row[4] else 'WAITING_TO_PRINT'
            })

        return jsonify({
            'success': True,
            'count': len(files),
            'files': files
        })

    except Exception as e:
        print(f"List files error: {e}")
        return jsonify({'error': True, 'message': 'Failed to list files'}), 500
    finally:
        cursor.close()
        release_db_connection(conn)

@files_bp.route('/print/<file_id>', methods=['GET'])
@token_required
def get_file_for_print(file_id):
    if g.user['role'] != 'owner':
        return jsonify({'error': 'Forbidden'}), 403

    conn = get_db_connection()
    cursor = conn.cursor()

    try:
        cursor.execute("""
            SELECT id, file_name, encrypted_file_data, file_size_bytes, 
                   iv_vector, auth_tag, encrypted_symmetric_key, created_at, 
                   is_printed, owner_id
            FROM files WHERE id = ? AND is_deleted = 0
        """, (file_id,))
        
        row = cursor.fetchone()
        if not row:
            return jsonify({'error': 'File not found'}), 404

        if str(row[9]) != g.user['sub']:
             return jsonify({'error': 'Forbidden: Not your file'}), 403

        # Convert bytes to Base64 strings for JSON response
        def to_b64(data):
            if isinstance(data, memoryview):
                data = data.tobytes()
            return base64.b64encode(data).decode('utf-8')

        response = {
            'success': True,
            'file_id': row[0],
            'file_name': row[1],
            'encrypted_file_data': to_b64(row[2]),
            'file_size_bytes': row[3],
            'iv_vector': to_b64(row[4]),
            'auth_tag': to_b64(row[5]),
            'encrypted_symmetric_key': to_b64(row[6]),
            'uploaded_at': row[7],
            'is_printed': row[8],
            'message': 'Decrypt this file on your PC before printing'
        }
        return jsonify(response)

    except Exception as e:
        print(f"Print download error: {e}")
        return jsonify({'error': True, 'message': 'Failed to download'}), 500
    finally:
        cursor.close()
        release_db_connection(conn)

@files_bp.route('/delete/<file_id>', methods=['POST'])
@token_required
def delete_file(file_id):
    if g.user['role'] != 'owner':
        return jsonify({'error': 'Forbidden'}), 403

    conn = get_db_connection()
    cursor = conn.cursor()

    try:
        # Verify ownership
        cursor.execute("SELECT owner_id, is_deleted FROM files WHERE id = ?", (file_id,))
        row = cursor.fetchone()
        if not row:
             return jsonify({'error': 'File not found'}), 404
        
        if str(row[0]) != g.user['sub']:
             return jsonify({'error': 'Forbidden'}), 403
        
        if row[1]:
             return jsonify({'error': 'File already deleted'}), 400

        # Mark as deleted
        deleted_at = datetime.datetime.utcnow().isoformat()
        cursor.execute("""
            UPDATE files 
            SET is_deleted = 1, deleted_at = ?, is_printed = 1, printed_at = ?
            WHERE id = ?
        """, (deleted_at, deleted_at, file_id))
        
        cursor.execute("SELECT id, file_name FROM files WHERE id = ?", (file_id,))
        deleted = cursor.fetchone()
        conn.commit()

        return jsonify({
            'success': True,
            'file_id': deleted[0],
            'file_name': deleted[1],
            'status': 'DELETED',
            'deleted_at': deleted_at,
            'message': 'File deleted'
        })

    except Exception as e:
        conn.rollback()
        print(f"Delete error: {e}")
        return jsonify({'error': True, 'message': 'Delete failed'}), 500
    finally:
        cursor.close()
        release_db_connection(conn)
